CNCF Zero Trust Whitepaper (#1229)
cncf/tag-security: zero-trust-whitepaper/v1/cloud-native-zero-trust-whitepaper.md

<!-- cspell:disable -->
* [Zero Trust (techskillnation.com)](https://techskillnation.com/2022/12/zero-trust/)

     <!-- cspell:enable -->Google was one of the first companies to take the cue from the potentialities of the Zero Trust architecture. The initiative was called BeyondCorp, and in 2009, Google implemented a de-parameterized framework based on the Zero Trust Architecture.
<!-- cspell:disable -->
* [Zero Trust Network Architecture the Best Fit for Your Use Cases (infraon.io)](https://infraon.io/blog/zero-trust-network-access-use-cases/)<!-- cspell:enable -->

    NIST played a significant role in popularizing and formalizing the zero trust security model

<!-- cspell:disable -->
* [Zero Trust Architecture (nvlpubs.nist.gov)](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-207.pdf) <!-- cspell:enable -->

    A zero trust architecture is designed and deployed with adherence to the following zero trust
basic tenets….

## Acknowledgments

We want to thank several contributors from whom we had excellent input and feedback and, as leading practitioners in the field, did much of the work that we write about in this document:
<!-- cspell:disable -->

> * José Carlos Chávez (Tetrate)

Reviewer: ```suggestion
* José Carlos Chávez (Okta)
```

### Client Identities

Identity verification can be based on either client-sent tokens, mTLS client certificates, or a combination of both. Typically, the identity of the workload or service is represented by an Identity Token, also known as an Access Token, included with every request. This token can be inspected by a chain of Policy Enforcement Points to control access. Certificates, on the other hand, are designed to encrypt the connection between two points and guarantee the identity of one (TLS) or both (mTLS) access points.

It's important to note that identity verification is only the first step in confirming the client’s identity. Verified identities under Zero Trust also need to be monitored by an _Active Observer_ to ensure that credentials are not being misused for malicious activity.

#### Token-Based Identity

> When using Tokens such as [JSON Web Tokens (JWT)](https://jwt.io) to establish identity, the client obtains a token from a trusted third party. The token is then sent as part of the service request, enabling the service to verify the client’s identity using the trusted third party. Tokens, preferably short-lived, are often used for authentication, allowing a service to verify a client’s identity even when the service request is sent via intermediate proxies. As the request with the Identity Token travels across the endpoints, the identity attributes can be easily read by Policy Enforcement Points, or a chain of them, allowing them to accept or deny access to various resources along the way while preserving the integrity of the token (it cannot be modified).

Reviewer 2: Should the link to JWT be RFC 7519?

```suggestion
When using Tokens such as [JSON Web Tokens (JWT)](https://datatracker.ietf.org/doc/rfc7519/) to establish identity, the client obtains a token from a trusted third party. The token is then sent as part of the service request, enabling the service to verify the client’s identity using the trusted third party. Tokens, preferably short-lived, are often used for authentication, allowing a service to verify a client’s identity even when the service request is sent via intermediate proxies. As the request with the Identity Token travels across the endpoints, the identity attributes can be easily read by Policy Enforcement Points, or a chain of them, allowing them to accept or deny access to various resources along the way while preserving the integrity of the token (it cannot be modified).
```

* 2022-2023 Best Undergraduate Cybersecurity Programs - US News Rankings: [https://www.usnews.com/best-colleges/rankings/computer-science/cybersecurity](https://www.usnews.com/best-colleges/rankings/computer-science/cybersecurity)
* Referenced Projects:
  * CNCF Knative’s Security-Guard: [https://knative.dev/docs/serving/app-security/security-guard-about/](https://knative.dev/docs/serving/app-security/security-guard-about/)
  * Falco: [https://falco.org/](https://falco.org/)
  * Cilium: [https://cilium.io/](https://cilium.io/)
  * Pixie: [https://docs.px.dev/](https://docs.px.dev/)
  * KubeArmor: [https://kubearmor.io/](https://kubearmor.io/)
  * Curiefense: [https://www.curiefense.io/](https://www.curiefense.io/)
  * Istio: [https://istio.io/](https://istio.io/)
  * Linkerd: [https://linkerd.io/](https://linkerd.io/)
  * Dapr: [https://dapr.io/](https://dapr.io/)
  * Knative: [https://knative.dev/](https://knative.dev/)
  * Dex: [https://dexidp.io/](https://dexidp.io/)
  * Keycloak: [https://www.keycloak.org/](https://www.keycloak.org/)
  * SPIFFE and SPIRE: [https://spiffe.io/](https://spiffe.io/)

>   * JSON Web Tokens (JWT): [https://jwt.io](https://jwt.io)

Reviewer 2: Should the link to JWT be RFC 7519?
```suggestion
  * JSON Web Tokens (JWT): [https://datatracker.ietf.org/doc/rfc7519/](https://datatracker.ietf.org/doc/rfc7519/)
```

  * Cert-manager: [https://cert-manager.io/](https://cert-manager.io/)
  * Kubernetes Network Policies: [https://kubernetes.io/docs/concepts/services-networking/network-policies/](https://kubernetes.io/docs/concepts/services-networking/network-policies/)

### History of Zero Trust:

> * [https://infraon.io/blog/history-of-zero-trust-security/](https://infraon.io/blog/history-of-zero-trust-security/) <!-- cspell:enable -->

Reviewer 2: Why not unify the label using the page title instead of the URL?
```suggestion
* [History of Zero Trust Security | Best Introduction 2022 | Infraon](https://infraon.io/blog/history-of-zero-trust-security/) <!-- cspell:enable -->
```

    In April 1994, the term "zero trust" was coined by Stephen Paul Marsh in his doctoral thesis on computer security at the University of Stirling. Marsh's work studied trust as something finite that can be described mathematically, asserting that the concept of trust transcends human factors such as morality, ethics, lawfulness, justice, and judgment.

> * [TR133.pdf (stir.ac.uk)](https://www.cs.stir.ac.uk/~kjt/techreps/pdf/TR133.pdf)

Reviewer 2: ```suggestion
* [Formalising Trust as a Computational Concept (stir.ac.uk)](https://www.cs.stir.ac.uk/~kjt/techreps/pdf/TR133.pdf)
```

    <!-- cspell:disable -->
    John Kindervag, a former analyst at Forrester Research, in 2010. John Kindervag <!-- cspell:enable --> published a research paper titled "No More Chewy Centers: Introducing the Zero Trust Model of Information Security," in which he introduced the concept of zero trust as a new approach to cybersecurity.

   <!-- cspell:disable -->

> * [Forrester-No-More-Chewy-Centers.pdf (crystaltechnologies.com)](https://crystaltechnologies.com/wp-content/uploads/2017/12/forrester-zero-trust-model-information-security.pdf)

Reviewer 2: ```suggestion
* [No More Chewy Centers: The Zero Trust Model Of Information Security (crystaltechnologies.com)](https://crystaltechnologies.com/wp-content/uploads/2017/12/forrester-zero-trust-model-information-security.pdf)
```

    The concept of Zero Trust was first introduced by
    John Kindervag, a former analyst at Forrester Research, <!-- cspell:enable --> in 2010. In his model, a network is divided into “zones” with varying levels of trust. The innermost zone, known as the “inner circle,” is the most trusted and contains the organization’s most sensitive data and resources. As you move outward from the inner circle, the level of trust decreases, with the outermost zone being the least trusted.

> * [https://techskillnation.com/2022/12/zero-trust/](https://techskillnation.com/2022/12/zero-trust/)

Reviewer 2: ```suggestion
* [Zero Trust (techskillnation.com)](https://techskillnation.com/2022/12/zero-trust/)
```

    Google was one of the first companies to take the cue from the potentialities of the Zero Trust architecture. The initiative was called BeyondCorp, and in 2009, Google implemented a de-parameterized framework based on the Zero Trust Architecture.
<!-- cspell:disable -->
* [Zero Trust Network Architecture the Best Fit for Your Use Cases | 2022 | Infraon](https://infraon.io/blog/zero-trust-network-access-use-cases/)<!-- cspell:enable -->

    NIST played a significant role in popularizing and formalizing the zero trust security model

> * [https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-207.pdf](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-207.pdf)

Reviewer 2: ```suggestion
* [Zero Trust Architecture (nvlpubs.nist.gov)](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-207.pdf)
```

We hope that the cloud native community, CNCF members, and other interested parties will use this whitepaper as a baseline for conversation and implementation of zero trust principles into their deployments while engaging in [feedback](https://github.com/cncf/tag-security/issues/950) to help evolve this paper.

## Principles of Zero Trust

### General Principles

Principles allow the reader to understand the fundamental ideas and beliefs that shape the rest of the paper's content. Principles serve as a reference point and help readers contextualize the information presented throughout the paper. They provide a logical and conceptual framework that supports the arguments, strategies, and recommendations discussed later in the document.

![alt_text](images/image2.png "image_tooltip")

_Image 2. In the context of Zero Trust, we model cloud native as an environment where “Service Request Senders” (users or machines from within the cloud or from outside the cloud) send “Service Requests” to Services implementing specific functions or features._

Principles of Zero Trust:

* **Assume a breach**.  Plan for systems and elements getting compromised.

>     1. **Assume every image includes vulnerabilities** - there is no such thing as a vulnerability-free image._ Dependencies are vulnerable, development tools get hacked, repositories get hacked, and CI/CD tools get hacked. The mountain of code we all use is vulnerable. _

Reviewer 2: ```suggestion
    1. **Assume every image includes vulnerabilities** - there is no such thing as a vulnerability-free image. _Dependencies are vulnerable, development tools get hacked, repositories get hacked, and CI/CD tools get hacked. The mountain of code we all use is vulnerable._
```

>     2. **Assume every service is vulnerable** - there is no such thing as a vulnerability-free service. _Services are vulnerable as they use images with vulnerabilities and exchange data that may be under an offender’s control. _

Reviewer 2: ```suggestion
    2. **Assume every service is vulnerable** - there is no such thing as a vulnerability-free service. _Services are vulnerable as they use images with vulnerabilities and exchange data that may be under an offender’s control._
```

    4. **Assume the cluster network is hostile and untrusted** - Assume injection and extraction of cluster network data. Never trust internal cluster traffic more than you trust ingress traffic. _Cluster networks do get hacked. Services on clusters do get hacked, and lateral movement is a thing._
    5. **Assume request senders (aka clients) may potentially send malicious requests** - Even when presenting valid credentials, a request may still be hostile. Credentials get stolen, clients get hacked, users may seek to abuse their access rights, etc.
* **Always verify, never trust.** This Zero Trust tenet translates into three aspects:
  * (a) **Eliminate implicit trust**
  * (b) **Continually monitor actual behavior to verify trustworthiness**
  * (c) **Minimize explicit trust**

    In the context of cloud native:
    * (a) **Eliminate implicit trust**
      * **Always authenticate the service** - internal and external clients need to verify the identity of any service approached.

>             * The internal cluster network can also be hostile; hence, even internal clients must authenticate the services and protect against Man-in-the-Middle attacks on the internal network.
>         * **Always authenticate service request senders** - verify the identity of any request sender (aka client), either cluster internal or external to any service. This includes both users as well as machines sending requests.
>           * We should assume the internal cluster network is also hostile, and offenders may send requests from within the internal cluster network.

Reviewer 2: ```suggestion
        * The internal cluster network can also be hostile; hence, even internal clients must authenticate the services and protect against Man-in-the-Middle attacks on the internal network.
      * **Always authenticate service request senders** - verify the identity of any request sender (aka client), either cluster internal or external to any service. This includes both users as well as machines sending requests.
        * We should assume the internal cluster network is also hostile, and offenders may send requests from within the internal cluster network.
```

1. Past behavior of the client - both its normal behavior and the past **confidence level** of its requests.
2. External information about the identity - e.g., When the identity is another service, such information may include the service behavior as discussed in “**Always monitor the service behavior**”

Always grant minimal permissions to service requests and permit the minimal required behaviors

Organizations should use dynamic and fine-grain access control to ensure that the sender whose identity is verified is allowed to perform the operation required in the given context. Such access control includes:

1. The operation required by the identity - is it one that the identity should be allowed to make?
1. The **confidence level** of the true identity of the sender
1. The **confidence level** of the request not being an exploit

> 1. The overall context - are we expecting requests at this time of day? From this ip-range? Etc.

Reviewer 2: ```suggestion
1. The overall context - are we expecting requests at this time of day? From this IP-range? Etc.
```

We can break this into three simple steps to keep in mind.

![alt_text](images/image3.png "image_tooltip")
_Image 3._

#### Always identify

No matter where the client is, whenever a client tries to access, validate that the client has proper credentials. Never trust a client based on other attributes, such as their location. Keep track of all clients based on their individual identities. Make sure to periodically revalidate the client credentials to detect clients with invalidated credentials. Avoid long sessions based on past credential validations.

#### Always analyze

Just because a client authentication was successful, it doesn’t mean the client is trustworthy. All service request traffic, whether internal or external, is treated as potentially malicious  - analyze it and verify it is not an attack. Insider threats and malicious actors may gain access to valid credentials. Legitimate clients with proper credentials may be compromised and act maliciously.

> Clients, client requests, and services need to be continuously analyzed and based on their behavior compared to an expected criteria. Continuously record all network and system activity and regularly analyze and inspect it to verify what occurs. Make sure you can replay decisions for audit purposes.

Reviewer 2: ```suggestion
Clients, client requests, and services need to be continuously analyzed and based on their behavior compared to expected criteria. Continuously record all network and system activity and regularly analyze and inspect it to verify what occurs. Make sure you can replay decisions for audit purposes.
```

### Cloud Native Zero Trust Core Elements

The [DoD Reference Architecture](https://dodcio.defense.gov/Portals/0/Documents/Library/(U)ZT_RA_v2.0(U)_Sep22.pdf) identifies seven core pillars: Users, Devices, Networking, Applications and Workloads, Data, and Automation and Orchestration. We identify the following as the Bedrocks of cloud native Zero Trust:

![alt_text](images/image4.png "image_tooltip")

_Image 4. Note that the service request sender (aka client) may be malicious. The Network and Data may also be compromised._

* **Service Instances** - Securing services offered using containers on cloud native clusters. DevSecOps practices to secure applications from inception till production. Secure by design practices, image build methodologies, image scanning, and storage. Service runtime protection methods include behavioral monitoring methods of services to form **confidence levels** in services.

> * **Clients Identities** - Securing the identity of clients sending service requests, either external or internal, monitoring their behavior to form a confidence** level** in identities.

Reviewer 2: ```suggestion
* **Clients Identities** - Securing the identity of clients sending service requests, either external or internal, monitoring their behavior to form a **confidence level** in identities.
```

* **Service requests** - Securing the requests sent by clients and processed by services. Monitoring the risks posed by requests to form **confidence levels** in risks from requests.
* **Data** - Categorization of data to enable securing data by data types.
* **Network** - Securing service requests and responses sent between Clients and Services, including encryption mechanisms.
* **Access Control** - Implementation of policies governing controls to secure access to services. Implementation of micro-segmentation in a cluster and enforcing granular access controls using gates in front of services. Considering the **confidence level** of identities and requests as part of access control. Considering Data types of both requests and responses as part of access control.
* **Automation** - Automation involves using tools and processes to automate the deployment, configuration, management, and audit of cloud native components while ensuring zero trust principles are followed. Automated removal of suspected service instances based on the service **confidence level**.

> We identify that **Client Identities** may be malicious, **Service Requests** may include exploits, **Service Instances** may be exploited, **Data** may include exploits, and the **Network** may be hostile. Since we can no longer trust such fundamental entities in our cloud native systems, we require an _“Active Observer”_ to continually assess the _“Confidence Level_”we have for such entities.

Reviewer 2: ```suggestion
We identify that **Client Identities** may be malicious, **Service Requests** may include exploits, **Service Instances** may be exploited, **Data** may include exploits, and the **Network** may be hostile. Since we can no longer trust such fundamental entities in our cloud native systems, we require an _“Active Observer”_ to continually assess the _“Confidence Level”_ we have for such entities.
```

_Image 5._

In the context of Cloud Services, we identify between:

* **Security Behavior Analytics for Service Requests (SBA-SR)** - the use of SBA for analyzing the behavior exposed by **service requests** before they arrive at a vulnerable service.
* **Security Behavior Analytics for Service Instances (SBA-SI)** - the use of SBA for analyzing the behavior exposed by cloud **service.**
* The behavior of **client identities** may be deduced from the behavior of requests emitted by such clients. When the client is a Cloud native service, its behavior is also deduced from the behavior of the service.

By applying zero trust principles to the Bedrocks of Zero Trust, we have a structured framework to operationalize the key components required when applying it to cloud native architectures to ensure comprehensive coverage across different dimensions of zero trust.

## Zero Trust Components

### Trust in a Zero Trust Cloud Native Platform

When applying zero trust principles to cloud native platforms, we need to consider what zero trust is, why we need it, the components of these platforms, and how our principles should be applied to each.

> The elephant in the room is that the "zero" in Zero Trust does not mean we trust nothing. All systems require “trust;” it is how that trust is obtained that changes. With zero trust, trust is not a given; it is not enough to be inside the cloud native platform to be trusted. Perimeter security is no longer enough; we must treat cloud native **clients**, **service instances**, **service requests**, **data**, and **network** as untrusted. Zero trust requires continual and dynamic assessment to evaluate if trust is to be given at any given moment to any given entity. We no longer trust “by design;” we trust after making a specific assessment given all the information we can collect at any given moment.

Reviewer 2: ```suggestion
The elephant in the room is that the "zero" in Zero Trust does not mean we trust nothing. All systems require “trust;” it is how that trust is obtained that changes. With zero trust, trust is not a given; it is not enough to be inside the cloud native platform to be trusted. Perimeter security is no longer enough; we must treat cloud native **clients**, **service instances**, **service requests**, **data**, and **networks** as untrusted. Zero trust requires continual and dynamic assessment to evaluate if trust is to be given at any given moment to any given entity. We no longer trust “by design;” we trust after making a specific assessment given all the information we can collect at any given moment.
```

#### Encryption

Zero trust assumes that offenders may reside inside cloud clusters, stage attacks on the cluster network, and attempt to access any data stored in the cloud. Therefore, all data, at rest and in transit, should be encrypted under Zero Trust. This protects sensitive data if it falls into the wrong hands and forms part of the process of authenticating and verifying interactions.

#### Container Security

Container security should follow security best practices by applying security measures at the container level. This includes secure configuration (ensuring containers are set up securely) and vulnerability scanning (checking for known vulnerabilities in your container images).

Additionally, under Zero Trust, runtime security is required to monitor suspicious activity while containers are running.

### The _Active Observer_

> Under cloud native Zero Trust, we require an _Active Observer _to establish trust. It is not enough to verify the **client’s** identity using Identity Management; for example, we need to consider that credentials may be stolen and the **client** may be compromised. An _Active Observer_ should be in place to observe that client’s activities and decide on the _Confidence Level _assigned to that identity - how sure are we that the **client** is not compromised and that the entity presenting the credentials is indeed the **client**?

Reviewer 2: ```suggestion
Under cloud native Zero Trust, we require an _Active Observer_ to establish trust. It is not enough to verify the **client’s** identity using Identity Management; for example, we need to consider that credentials may be stolen and the **client** may be compromised. An _Active Observer_ should be in place to observe that client’s activities and decide on the _Confidence Level_ assigned to that identity - how sure are we that the **client** is not compromised and that the entity presenting the credentials is indeed the **client**?
```

Similarly, an _Active Observer_ is required to verify that the client is sending legitimate **service requests** rather than malicious ones. And that the **Service Instances** processing such requests are not compromised.

#### Continuous Monitoring and Logging

> Continuous monitoring is crucial for a Zero Trust approach. This involves collecting and analyzing logs to detect unusual activity and setting up alerts for possible security incidents. Security monitoring plays an important role as an input to an _Active Observer_. It may be used to detect compromised **Clients** and **Service Instances.

Reviewer 2: ```suggestion
Continuous monitoring is crucial for a Zero Trust approach. This involves collecting and analyzing logs to detect unusual activity and setting up alerts for possible security incidents. Security monitoring plays an important role as an input to an _Active Observer_. It may be used to detect compromised **Clients** and **Service Instances**.
```

Detecting such irregularities is a key element in detecting exploited cloud service instances. The deployed SBA-SI should be capable of discerning regular service behavior intended by the organization from the composite behavior affected by the malware. The outcome of this component is a _Confidence Level_ of a **Service Instance** that should be handled by **Automation**, as discussed below.

### Access Control

Under Zero Trust, **Access Control** takes new roles in several dimensions:

* Access control should be applied for specific **client identities** rather than for “all clients” or groups of clients.
* Access control should be applied per **Service** rather than for “all services” or groups of services. All Service Instances should naturally have the same access control.
* Access control should be considered per **Service Request** rather than being applied once for a client and applied for all the client requests.
  * When access control is considered for a **Service Request**, it should consider the following:

>     * The _Confidence Level_ of the** Service Requests **(how sure are we that this is not an exploit or part of an attack)
>     * The _Confidence Level_ of the** Client identity **(how sure are we that this client is not compromised)

Reviewer 2: ```suggestion
    * The _Confidence Level_ of the **Service Requests** (how sure are we that this is not an exploit or part of an attack)
    * The _Confidence Level_ of the **Client identity** (how sure are we that this client is not compromised)
```

Combined, Zero Trust **Access Control** apply

1. The principle of least privilege (PoLP) gives a **client** connecting to a **service** only the privileges the client needs to perform their job as part of the service and nothing more. This can be achieved by a Role-Based Access Control (RBAC)  that will be correctly used to ensure each client is authorized based on its identity and the identity of a specific cloud service being approached. Other options include ABAC (Attribute-Based Access Control) and ReBAC (Relationship-Based Access Control).

> 2. Dynamic and context-aware access control, taking real-time access decisions on a per-request basis. Decisions are based on the _Confidence Levels _evaluated by an _Active Observer_. This can be achieved by introducing SBA technology and enhancing Access Control to enable the introduction of dynamic access rules and policies.

Reviewer 2: ```suggestion
2. Dynamic and context-aware access control, taking real-time access decisions on a per-request basis. Decisions are based on the _Confidence Levels_ evaluated by an _Active Observer_. This can be achieved by introducing SBA technology and enhancing Access Control to enable the introduction of dynamic access rules and policies.
```

| Client Identities | Verify the identity of every service request sender (client). Use <em>Active Observer</em> to monitor and verify client behavior over time. Flag or auto-block suspected clients.|
| Access Control | Use fine-grained access control in front of every service. Control which client identities can send service requests while considering their confidence level. Control which service requests these clients are allowed to send while considering the service requests' confidence level.|
-----

_A summary of the best practice topics to support ZTA principles using CNCF technologies_

### Service Instances

**Service instances** under CNCF Zero Trust need to be monitored using an _Active Observer_ to achieve two goals:

* Identifying compromised instances that can be stopped and  non-compromised instances started to replace them.
* Helping to identify compromised client identities when the affected service instance is also a client of other cloud services.
* Allowing to run forensic analysis to identify the kind of ongoing attack

#### Security Behavior Analytics for Service Instances (SBA-SI)

> CNCF Knative’s [Security-Guard](https://knative.dev/docs/serving/app-security/security-guard-about/) (Guard) offers support as an SBA-SI. Like other Knative components, Guard can be used separately from Knative and be deployed as part of vanilla Kubernetes or other cloud native orchestration systems. Guard uses machine-learning-based criteria synthesis to identify the standard patterns service clients use. Guard also supports setting manual criteria. Guard’s SBA-SI identifies changes in the external communication performed by service instances. It calculates a **confidence level** of **service instances** and is integrated with **Automation**,** **which optionally allows deleting service instances suspected as being exploited.

Reviewer 2: ```suggestion
CNCF Knative’s [Security-Guard](https://knative.dev/docs/serving/app-security/security-guard-about/) (Guard) offers support as an SBA-SI. Like other Knative components, Guard can be used separately from Knative and be deployed as part of vanilla Kubernetes or other cloud native orchestration systems. Guard uses machine-learning-based criteria synthesis to identify the standard patterns service clients use. Guard also supports setting manual criteria. Guard’s SBA-SI identifies changes in the external communication performed by service instances. It calculates a **confidence level** of **service instances** and is integrated with **Automation**, which optionally allows deleting service instances suspected as being exploited.
```

#### Token-Based Identity

When using Tokens such as [JSON Web Tokens (JWT)](https://jwt.io) to establish identity, the client obtains a token from a trusted third party. The token is then sent as part of the service request, enabling the service to verify the client’s identity using the trusted third party. Tokens, preferably short-lived, are often used for authentication, allowing a service to verify a client’s identity even when the service request is sent via intermediate proxies. As the request with the Identity Token travels across the endpoints, the identity attributes can be easily read by Policy Enforcement Points, or a chain of them, allowing them to accept or deny access to various resources along the way while preserving the integrity of the token (it cannot be modified).

In adherence to current best practices for utilizing JWT Tokens, it is recommended to perform the following checks:

* **Signature**: Verify if the token has not been tampered with.
* **exp** and **nbf** Claims: Validate the expiration time of the token.
* **iss** Claim: Confirm the issuer's identity.
* **aud** Claim: Check if the audience includes the workload itself.

> * **Scope Claim** (if exists): For example, ensure that the resource owner authorizes access to the resource.

Reviewer 2: ```suggestion
* **scope** Claim (if exists): For example, ensure that the resource owner authorizes access to the resource.
```

#### Certificate-Based Identity

Identity may alternatively be established by the client presenting a client certificate using mutual TLS (mTLS). mTLS allows bi-directional, certificate-based verification. The client can verify the Server, and the Server can verify the client. mTLS uses certificates for hop-by-hop identity verification, which allows a service to verify the identity of a directly connected client co-located on the cloud cluster network. CNCF projects for managing certificates include [cert-manager](https://cert-manager.io/) - cloud native certificate management.

Making mTLS standard for all communications in a cloud cluster is often achieved by deploying a service mesh. CNCF Service mesh projects [Istio](https://istio.io/), [SPIFFE](https://spiffe.io/), and [Linkerd](https://linkerd.io/) offer service-to-service identity verification mTLS.  [Dapr](https://dapr.io/) also offers mTLS.

#### _Active Observer_ for Client Identities

> **Confidence level** of local clients embedded in cloud clusters may be evaluated by aggregating the **confidence level** of the respective** service instances** that embed such local clients. Inputs about external clients may or may not be obtainable.

Reviewer 2: ```suggestion
**Confidence level** of local clients embedded in cloud clusters may be evaluated by aggregating the **confidence level** of the respective **service instances** that embed such local clients. Inputs about external clients may or may not be obtainable.
```

#### Fine-grained Access Control

Service mesh based solutions ([Istio](https://istio.io/), [Linkerd](https://linkerd.io/)) offer opportunities to implement a gate in front of every service instance and support fine-grained access policies. Alternatively, when using [Dapr](https://dapr.io/), you may use the Dapr access control.

#### Dynamic access control

> [Security-Guard](https://knative.dev/docs/serving/app-security/security-guard-about/) implements dynamic access control by introducing an Access Control gate in front of every **Service Instance**.The Access Control can be configured to block **service requests** based on the request **confidence level**.

Reviewer 2: ```suggestion
[Security-Guard](https://knative.dev/docs/serving/app-security/security-guard-about/) implements dynamic access control by introducing an Access Control gate in front of every **Service Instance**. The Access Control can be configured to block **service requests** based on the request **confidence level**.
```

## References

<!-- cspell:disable -->
* NIST SP 800-207, Zero Trust Architecture: [https://csrc.nist.gov/publications/detail/sp/800-207/final](https://csrc.nist.gov/publications/detail/sp/800-207/final)

> * NIST SP 800-207A, A Zero Trust Architecture Model for Access Control in Cloud-Native Applications in Multi-Cloud Environments:[https://csrc.nist.gov/pubs/sp/800/207/a/final](https://csrc.nist.gov/pubs/sp/800/207/a/final)

Reviewer 2: ```suggestion
* NIST SP 800-207A, A Zero Trust Architecture Model for Access Control in Cloud-Native Applications in Multi-Cloud Environments: [https://csrc.nist.gov/pubs/sp/800/207/a/final](https://csrc.nist.gov/pubs/sp/800/207/a/final)
```

* [https://infraon.io/blog/history-of-zero-trust-security/](https://infraon.io/blog/history-of-zero-trust-security/) <!-- cspell:enable -->

    In April 1994, the term "zero trust" was coined by Stephen Paul Marsh in his doctoral thesis on computer security at the University of Stirling. Marsh's work studied trust as something finite that can be described mathematically, asserting that the concept of trust transcends human factors such as morality, ethics, lawfulness, justice, and judgment.

* [TR133.pdf (stir.ac.uk)](https://www.cs.stir.ac.uk/~kjt/techreps/pdf/TR133.pdf)

    <!-- cspell:disable -->
    John Kindervag, a former analyst at Forrester Research, in 2010. John Kindervag <!-- cspell:enable --> published a research paper titled "No More Chewy Centers: Introducing the Zero Trust Model of Information Security," in which he introduced the concept of zero trust as a new approach to cybersecurity.

   <!-- cspell:disable -->
* [Forrester-No-More-Chewy-Centers.pdf (crystaltechnologies.com)](https://crystaltechnologies.com/wp-content/uploads/2017/12/forrester-zero-trust-model-information-security.pdf)

    The concept of Zero Trust was first introduced by

>     John Kindervag, a former analyst at Forrester Research,
>     <!-- cspell:enable -->
>     in 2010. In his model, a network is divided into “zones” with varying levels of trust. The innermost zone, known as the “inner circle,” is the most trusted and contains the organization’s most sensitive data and resources. As you move outward from the inner circle, the level of trust decreases, with the outermost zone being the least trusted.

Reviewer 2: ```suggestion
    John Kindervag, a former analyst at Forrester Research, <!-- cspell:enable --> in 2010. In his model, a network is divided into “zones” with varying levels of trust. The innermost zone, known as the “inner circle,” is the most trusted and contains the organization’s most sensitive data and resources. As you move outward from the inner circle, the level of trust decreases, with the outermost zone being the least trusted.
```

    NIST played a significant role in popularizing and formalizing the zero trust security model

* [https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-207.pdf](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-207.pdf)

    A zero trust architecture is designed and deployed with adherence to the following zero trust
basic tenets….

## Acknowledgments

We want to thank several contributors from whom we had excellent input and feedback and, as leading practitioners in the field, did much of the work that we write about in this document:
<!-- cspell:disable -->
* José Carlos Chávez (Tetrate)
* Matt Flannery
* Sundar Nadathur (Intel NEX)
* Andrew Martin
* Matteo Bisi
* Fabian Kammel
* Yaron Schneider
* Yoshiyuki Tabata
<!-- cspell:enable -->

For anyone involved in creating this paper, we express our heartfelt gratitude for your time, dedication, and professionalism. Your contributions have undoubtedly elevated the standard of our research, and we are immensely grateful for your collaboration.

> Thank you for being an integral part of this endeavor

Reviewer 2: ```suggestion
Thank you for being an integral part of this endeavor.
```

## Footnotes and Citations

[^1]:
    Internet Crime Complaint Center (IC3) Annual report:  [https://www.ic3.gov/Media/PDF/AnnualReport/2022_IC3Report.pdf](https://www.ic3.gov/Media/PDF/AnnualReport/2022_IC3Report.pdf)
[^2]:
    President Biden’s Zero Trust Mandate: [https://www.whitehouse.gov/wp-content/uploads/2022/01/M-22-09.pdf](https://www.whitehouse.gov/wp-content/uploads/2022/01/M-22-09.pdf)
[^3]:
    Department of Defense (DoD) Zero Trust Reference Architecture: [https://dodcio.defense.gov/Portals/0/Documents/Library/(U)ZT_RA_v2.0(U)_Sep22.pdf](https://dodcio.defense.gov/Portals/0/Documents/Library/(U)ZT_RA_v2.0(U)_Sep22.pdf)
[^4]:

>     Dorothy Denning- A New Paradigm for Trusted Systems, 1993  [https://dl.acm.org/doi/pdf/10.1145/283751.283772](https://dl.acm.org/doi/pdf/10.1145/283751.283772)

Reviewer 2: ```suggestion
    Dorothy Denning- A New Paradigm for Trusted Systems, 1993: [https://dl.acm.org/doi/pdf/10.1145/283751.283772](https://dl.acm.org/doi/pdf/10.1145/283751.283772)
```

Guard’s SBA-SR offers the added benefit of being able to detect unknown exploits to unknown vulnerabilities without the use of signatures. It can help protect services without the usual race between CVEs, Exploits, and Patches.

SBA-SR may also be embedded in Web Application Firewall (WAF) that maintains per service state and is capable of analyzing incoming/ongoing traffic and detecting threads and ongoing attacks. Solutions like [Coraza](https://coraza.io) or [Curiefense](https://www.curiefense.io/) and many others provide anomaly scores for requests and decide whether to block them based on predefined thresholds. Such anomaly scores can evolve to offer SBA-SR features.

#### Data in Transit

To ensure all communications in-transit are encrypted, all services should be using TLS (at least a version 1.2), and all clients need to verify the certificates presented by the server. CNCF projects offering TLS and certificates to protect inter-service communication include [Istio](https://istio.io/), [Linkerd](https://linkerd.io/), [Dapr](https://dapr.io/), and [Knative](https://knative.dev/).

> ### Client Identities

Reviewer 2: I will post the comment here that I posted in the Google document, too.
Shouldn't we clarify that token-based identity is recommended over certificate-based identity?
The Summary of Principles of Zero Trust states, "Authentication and authorization are handled on a per-request basis rather than on a per-session basis." I think it can be read that JWT (access token) is per request and mTLS is per session (TLS session).
In addition, in the subsequent access control (authorization) step, without a token, the scope authorized by the resource owner cannot be known (against the principle of least privilege), the audience cannot be known (token redirect attack exposed), and the transaction cannot be known (benefits such as Transaction Tokens, which have been recently discussed in the OAuth WG, are not be enjoyed).
I understand that there are entities that cannot adopt token-based identity and have no choice but to adopt certificate-based identity, but I think certificate-based identity has a lower security level than token-based identity, so we should recommend token-based identity. What do you think?

Author: Thank you @y-tabata for your comment. Apologies for not addressing it earlier, but the migration to Markdown has some challenges and I might have omitted it in the process. 
I understand your suggestion, and I will try to incorporate it to the text. I think we can position it as follow: a recommended approach is to use token, e.g. JWT,  representing workload identity, for a chain of Policy Enforcement Points, that might inspect the request, and allow/deny access to the resource based on the identity of the request (short-lived identity preferred). JWT, with the signature guarantees the integrity of the identity token representing the workload request, while it is easily readable all along the way. TLS, on the on the other hand, is designed to encrypt the connection between two points and guarantees the identity of one (TLS), or both (mTLS) access points.  

What do you think?

Reviewer 2: @mrsabath 
Thanks for addressing my comment. LGTM.
Further, when using JWT, I thought it would be helpful also to mention how each workload should use JWT.
In terms of current best practices, I think it is basic to check the following first.
- signature (check not tampered)
- exp claim and nbf claim (check the expiration time)
- iss claim (check the issuer)
- aud claim (check it includes the workload itself)
- (if exists) scope claim (for example check whether the resource owner authorizes to access the resource)

I agree that JWT is recommended, but as you mentioned, JWT and mTLS have different guarantee points and layers, so I thought it would be a good idea from a zero-trust perspective to show the reader that they also have the option of choosing both instead of having them choose between JWT and mTLS.

#### Certificate-Based Identity

Identity may alternatively be established by the client presenting a client certificate using mutual TLS (mTLS). mTLS allows bi-directional, certificate-based verification. The client can verify the Server, and the Server can verify the client. mTLS uses certificates for hop-by-hop identity verification, which allows a service to verify the identity of a directly connected client co-located on the cloud cluster network. CNCF projects for managing certificates include [cert-manager](https://cert-manager.io/) - cloud native certificate management.

Making mTLS standard for all communications in a cloud cluster is often achieved by deploying a service mesh. CNCF Service mesh projects [Istio](https://istio.io/), [SPIFFE](https://spiffe.io/), and [Linkerd](https://linkerd.io/) offer service-to-service identity verification mTLS.  [Dapr](https://dapr.io/) also offers mTLS.

#### _Active Observer_ for Client Identities

> **Confidence level** of local clients embedded in cloud clusters may be evaluated by aggregating the **confidence level** of the respective** service instances** that embed such local clients. Inputs about external clients may or may be obtainable.

Reviewer 2: may or may not?

Reviewer 3: ```suggestion
**Confidence level** of local clients embedded in cloud clusters may be evaluated by aggregating the **confidence level** of the respective** service instances** that embed such local clients. Inputs about external clients may or may not be obtainable.
```
